src/GSoft.Extensions.Mongo.ApplicationInsights/ApplicationInsightsEventSubscriber.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Net;
+using GSoft.Extensions.Mongo.Telemetry;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using MongoDB.Driver.Core.Events;
+
+namespace GSoft.Extensions.Mongo.ApplicationInsights;
+
+internal sealed class ApplicationInsightsEventSubscriber : AggregatorEventSubscriber
+{
+    private readonly TelemetryClient _telemetryClient;
+    private readonly MongoClientOptions _options;
+    private readonly ConcurrentDictionary<int, ActivityAwareOperationHolder> _currentOperationsMap;
+
+    public ApplicationInsightsEventSubscriber(TelemetryClient telemetryClient, MongoClientOptions options)
+    {
+        this._telemetryClient = telemetryClient;
+        this._options = options;
+        this._currentOperationsMap = new ConcurrentDictionary<int, ActivityAwareOperationHolder>();
+
+        this.Subscribe<CommandStartedEvent>(this.CommandStartedEventHandler);
+        this.Subscribe<CommandSucceededEvent>(this.CommandSucceededEventHandler);
+        this.Subscribe<CommandFailedEvent>(this.CommandFailedEventHandler);
+    }
+
+    private void CommandStartedEventHandler(CommandStartedEvent evt)
+    {
+        if (this._options.Telemetry.IgnoredCommandNames.Contains(evt.CommandName))
+        {
+            return;
+        }
+
+        var name = evt.GetCollectionName() is { } collectionName
+            ? $"{evt.DatabaseNamespace.DatabaseName}.{collectionName}.{evt.CommandName}"
+            : $"{evt.DatabaseNamespace.DatabaseName}.{evt.CommandName}";
+
+        var operation = this.StartActivityAwareDependencyOperation(name);
+
+        operation.Telemetry.Type = "mongodb";
+
+        operation.Telemetry.Target = evt.ConnectionId?.ServerId?.EndPoint switch
+        {
+            IPEndPoint endPoint => endPoint.Address + ":" + endPoint.Port,
+            DnsEndPoint endPoint => endPoint.Host + ":" + endPoint.Port,
+            _ => "unknown",
+        };
+
+        if (this._options.Telemetry.CaptureCommandText)
+        {
+            operation.Telemetry.Data = evt.Command.ToString();
+        }
+
+        // Originating activity must be captured AFTER that the operation is created
+        // Because ApplicationInsights SDK creates another intermediate Activity
+        var originatingActivity = Activity.Current;
+
+        this._currentOperationsMap.TryAdd(evt.RequestId, new ActivityAwareOperationHolder(operation, originatingActivity));
```

Comment:
`TryAdd` will never returns false as `evt.RequestId` is guaranteed to be unique (it uses `Interlocked.Increment`)